### DISCUSSION TITLE
Allow Multiple Selection in Tree View

### DISCUSSION THREAD

**Discussion by crawr** [zowe/zowe-explorer-vscode#1130] [Ideas]
Related to PR #1128 

Research: [https://ent.box.com/s/nhohq2edjxw1qa703rg69zr8zh1ftevw](https://ent.box.com/s/nhohq2edjxw1qa703rg69zr8zh1ftevw)

Open Questions:
1.	Should there be a separate context menu for actions with multiple selection? What about the inline icon, should it be removed?
2.	If the action produces a notification, it will show up multiple times. Should these notifications be routed to just the console to prevent multiple pop-ups?
3.	Will this be a pure UI implementation or should this be implemented with API for extenders?
4.	Aside from copying and deleting multiple PDS members, what other commands should be prioritized?

**Comment by dkelosky** [comment]
Very cool! Does this work if you select multiple and delete with the delete keyboard key?

**Reply by Alexandru-Dumitru** [reply]
In theory it should (I mean, it works already in File Explorer), but I think it matters to what function the DEL key is bound to, and I assume, in this case it should be mapped to the dedicated `delete` action for each node type (deleting a profile or a member is a different action).

**Reply by jellypuno** [reply]
Can we force that bounding of the DEL key? Like, for ZE explorer it will be bounded to delete member regardless if it used in the keyBindings definition?

**Reply by Alexandru-Dumitru** [reply]
We should be able to. From what I can understand, we just need to provide something like this in package.json:
```
{ "key": "delete",                "command": "moveFileToTrash",
                                     "when": "explorerResourceMoveableToTrash && explorerViewletVisible && filesExplorerFocus && !explorerResourceReadonly && !inputFocus" }, 
```
- this is the action for deleting files in File Explorer


https://code.visualstudio.com/api/references/vscode-api#commands

**Reply by crawr** [reply]
Yes, we can contribute keybindings per command. It will still be subject to precedence based on user preferences. So we can use either `Delete` or `Ctrl+Delete`. We can even define it both for Windows and Mac.

**Reply by dkelosky** [reply]
Thanks all for chiming in.  Using the delete key is my preference, even if I have to configure it myself, since that's how I delete within File Explorer view.

**Comment by jellypuno** [comment]
Another question: Can we select multiple items from different profiles (e.g. delete multiple jobs from different profiles)

**Reply by Alexandru-Dumitru** [reply]
yes, from selection perspective, it is possible.

**Reply by crawr** [reply]
theoretically, it will work. I will need to check which session it will use to delete which job. Although if you are able to submit the job, you should have the permission to delete it as well.

**Reply by Alexandru-Dumitru** [reply]
Would it make sense for each node type to carry it's own delete functionality? and just call it from the function (i.e. vs code command) which is bound to the Delete key or context menu / inline menu items? Kind of like saying "delete yourself", this way we don't need to take care of each context for each node.

**Comment by venkatzhub** [comment]
My take: 
1. No separate action menu item is required for multi select. As for  inline icon, can we detect a multi-select and remove for those cases ? 
2. Routing to the console is probably the best option  in this case, and may be one pop up to say that to the user (that messages are being directed to console ) would be good. 
3. What use cases do we envision for extenders? Any known cases or examples? If we believe there are cases or opportunities for extenders, then pushing it to the API would be good. 
4. From a multi select point of view, the multiple member delete and  copy are the highest priority. Nothing major jumps to my mind.

**Comment by katelynienaber** [comment]
Thoughts:
1. What if we had a separate context menu which only contained the supported multiple select options, when multiple items are selected? Ex. if you have 3 jobs selected, you only see the actions that are possible when multiple jobs are selected. We could slowly add support for all the functionality that is currently possible for single jobs, until the context menus are the same...and also we could add extra items to the multi-select menu if necessary.
For the second part of this question...imho button should not be removed and will only work for the single item. It's a bit nonintuitive to do it this way, but it isn't going to do anything the user doesn't want (ie. delete something they didn't select)...they can just try it, and only one job is deleted, and they will realize they should use the context menu. I guess the button can't be removed in VSCode dynamically, when someone has many items selected, and added back after they click out of the selected items.
2. Maybe we can make 1 notification for starting the delete action, and each time a deletion is successful it sends a message to the console?
Also idk if you checked this, but what happens when one of the items can't be deleted? Does it keep trying to delete the other items, or does it quit? Would be good if it tries to delete all of the items, and sends a success/failure notif to the console for each item, along with an error for the ones that failed.
Idk if this was what you were talking about with the unsupported array implementation in Explorer, but we could show a notification only once per array of items to delete...and write something to the console for each array element.
3. Not sure about question 3 personally
4. I think the "ultimate adaptation" of multi-select would be, if you could select multiple of ANYTHING in the whole sidebar, from any tree or session, and delete it in one go...ie. select multiple sessions, multiple jobs, multiple USS folders/files, multiple data sets/members, and delete all of them at once...then people could do "spring cleaning". But I think that maybe this is overkill. If someone tries to delete multiple items from across trees, we could simply show an error. We could also disallow selecting multiple TYPES of items with an error, and add it later...so if someone selects a session, and a job, and tries to delete both, they aren't able to until it is supported.

Brainstorms for all possible multiple-processing possibilities that I can think of now:
_All of these could be done for multiple users, ie. selecting members/jobs/files belonging to 2 or more users_
_I have no idea which of these are really useful this is just BRAINSTORMING to list all possible options_
1. Data set tree:
- Multiple member select, for job submission?
- Multiple select for copy/paste (could also multiple copy/paste member or file content as text, with some formatting to separate contents of multiple files)
2. USS Tree:
- Multiple select for copy/paste
3. Jobs tree
- Multiple job select for downloading spools
- Multiple job select for Issue stop command
4. All trees:
- Multiple select for Add to favorites (this could also be cross-tree, ie. from Jobs & Data Sets at the same time, or whatever)
- Multiple select for Remove from favorites
- Multiple select for hide/delete session

**Comment by jellypuno** [comment]
- Delete only selected items (e.g do not delete the PDS if only a few members are selected)

- If a PDS has 10 members and the user selected the parent PDS and 3 members then the operation should prevent deleting the parent PDS.

- Define the use cases [ action item @jellypuno ]

- How the deletion is reported? Standardization on the reporting or dialog box between dataset and jobs

- What happens with error handling? e.g. 1st member didn't delete should it continue? We must standardize and document

- [ Investigated by @JillieBeanSim ] Check the issue with the mouse click on VSCode. The user tried to delete a member through right-click but (maybe) the PDS was highlighted so the PDS was deleted instead of the member

**Comment by jellypuno** [comment]
- [ Needs Research ]Can we prevent the right-click when deleting a dataset? maybe we can only allow deletion through a command

- [ Needs Research ] Can we update the tree to identify the datasets/member that I want to delete? For example, I selected several datasets then I hit delete. After that the tree will be updated with an icon or greyed out text to tell me that these are the datasets that I want to delete. At the same there will be a pop-up confirming if I want to continue with the deletion of the datasets. If yes, then those datasets are deleted and removed from the tree

**Comment by crawr** [comment]
Additional item to be considered: making the user experience consistent for functions that handle multi-select.
For example, the jobs deletion should also implement the progress bar, similar to the datasets deletion.
Related issue: #1338

**Comment by lauren-li** [comment]
Regarding what can be done with the UX and the related issue #1470, it seems that in VS Code’s own file explorer, if you left click a folder > right-click a child file and select “Delete”, it seems to register that the child file is the thing to delete, even though it’s the parent folder that remains highlighted. Perhaps that means we can implement something similar for our delete functionality to resolve #1470.

<img width="688" alt="Screen Shot 2021-09-09 at 1 49 48 PM" src="https://user-images.githubusercontent.com/45975633/132737814-b8024576-7a39-45f5-a8d0-c0446578fe07.png">
